modules/integration/tests-common/jacoco-report-generator/src/main/java/org/wso2/carbon/identity/jacoco/CodeCoverageUtils.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.jacoco;
+
+import org.codehaus.plexus.util.DirectoryScanner;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+/**
+ * Utility class for code coverage related operations.
+ */
+public class CodeCoverageUtils {
+
+    private static final int BUFFER_SIZE = 1024;
+    private static final String INVALID_EXTENSION_ERROR = "Invalid extension: %s is invalid";
+    private static final String EXTRACTION_ERROR = "Error on archive extraction";
+
+    /**
+     * Extract jar files given at jar file path
+     *
+     * @param jarFilePath - Jar file path
+     * @param tempDir     - Temporary directory to extract jar file
+     * @return - Jar file extracted directory.
+     * @throws IOException - Throws if jar extraction fails
+     */
+    public synchronized static String extractJarFile(String jarFilePath, File tempDir) throws IOException {
```

Comment:
Do we really need this to be `synchronized` here?

Reply:
Done with https://github.com/wso2/product-is/pull/21108